### ISSUE TITLE
Query plans not deterministic

### ISSUE THREAD

**Issue by ArturGajowy** [prestodb/presto#6579]
There are cases where subsequent executions of a query differ because of the query plan being created in a non-deterministic way. The simplest example we've spotted so far is:

```
                SELECT COUNT(*) FROM part ps, supplier l
                WHERE ps.comment = l.comment
                AND ps.brand = l.address
```

which produces at least two different plans:

```
- Output[_col0] => [count:bigint]
        _col0 := count
    - Aggregate(FINAL) => [count:bigint]
            count := "count"("count_10")
        - LocalExchange[SINGLE] () => count_10:bigint
            - Aggregate(PARTIAL) => [count_11:bigint]
                    count_11 := "count"(*)
                - Project => []
                    - InnerJoin[("comment" = "comment_1") AND ("brand" = "address")][$hashvalue, $hashvalue_12] => [brand:varchar(10), comment:varchar(23), $hashvalue:bigint, comment_1:varchar(101), address:varchar(40), $hashvalue_12:bigint]
                        - Project => [brand:varchar(10), comment:varchar(23), $hashvalue:bigint]
                                $hashvalue := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("comment"), 0)), COALESCE("$operator$hash_code"("brand"), 0))
                            - TableScan[local:tpch:part:sf0.01, originalConstraint = true] => [brand:varchar(10), comment:varchar(23)]
                                    brand := tpch:brand
                                    comment := tpch:comment
                        - LocalExchange[HASH][$hashvalue_12] ("comment_1", "address") => comment_1:varchar(101), address:varchar(40), $hashvalue_12:bigint
                            - Project => [comment_1:varchar(101), address:varchar(40), $hashvalue_13:bigint]
                                    $hashvalue_13 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("comment_1"), 0)), COALESCE("$operator$hash_code"("address"), 0))
                                - TableScan[local:tpch:supplier:sf0.01, originalConstraint = true] => [address:varchar(40), comment_1:varchar(101)]
                                        address := tpch:address
                                        comment_1 := tpch:comment
```
and
```
- Output[_col0] => [count:bigint]
        _col0 := count
    - Aggregate(FINAL) => [count:bigint]
            count := "count"("count_14")
        - LocalExchange[SINGLE] () => count_14:bigint
            - Aggregate(PARTIAL) => [count_15:bigint]
                    count_15 := "count"(*)
                - Project => []
                    - InnerJoin[("brand" = "address") AND ("comment" = "comment_1")][$hashvalue, $hashvalue_16] => [brand:varchar(10), comment:varchar(23), $hashvalue:bigint, comment_1:varchar(101), address:varchar(40), $hashvalue_16:bigint]
                        - Project => [brand:varchar(10), comment:varchar(23), $hashvalue:bigint]
                                $hashvalue := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("brand"), 0)), COALESCE("$operator$hash_code"("comment"), 0))
                            - TableScan[local:tpch:part:sf0.01, originalConstraint = true] => [brand:varchar(10), comment:varchar(23)]
                                    brand := tpch:brand
                                    comment := tpch:comment
                        - LocalExchange[HASH][$hashvalue_16] ("address", "comment_1") => comment_1:varchar(101), address:varchar(40), $hashvalue_16:bigint
                            - Project => [comment_1:varchar(101), address:varchar(40), $hashvalue_17:bigint]
                                    $hashvalue_17 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("address"), 0)), COALESCE("$operator$hash_code"("comment_1"), 0))
                                - TableScan[local:tpch:supplier:sf0.01, originalConstraint = true] => [address:varchar(40), comment_1:varchar(101)]
                                        address := tpch:address
                                        comment_1 := tpch:comment
```

Such a plan difference, in an isomorphic query, resulted in an order of magnitude performance difference (because of the table sizes). In other queries, the plan differences were more substantial, like computing additional hashvalues (which - in case of deterministic planning - would have been duplicates and thus eliminated by the optimizer) or even changing the join tree structure in one of TPC-DS queries.

Fixes and test cases for above issues are in flight here:
https://github.com/Teradata/presto/compare/master...deterministic_query_plans

We'd like to automatically test plan stability for all the TPC-H and TPC-DS, and possibly AbstractTestQueries. The last one takes about 50s when tested for plan determinism only.

One of the nondeterminism sources (and the way of its eradication) needs some discussion, so I'm going to open a draft PR so we can discuss it there.

**Comment by ArturGajowy** [comment]
cc @martint

**Comment by ArturGajowy** [comment]
fixed by #6771
